src/_pytest/assertion/util.py

```diff
@@ -147,7 +147,9 @@ def assertrepr_compare(config, op, left, right):
     explanation = None
     try:
         if op == "==":
-            if istext(left) and istext(right):
+            if isinstance(left, bytes) and isinstance(right, bytes):
+                explanation = _diff_text(left, right, verbose)
+            elif istext(left) and istext(right):
```

Comment:
Note that this compares bytes and unicode on py2 still..

Reply:
Maybe this as a follow up?
But it changes `test_nonascii_text`, since the type is the derived class then (https://github.com/pytest-dev/pytest/pull/5268/files#diff-14ad89cac128f1dd3aaa927c656035b3R597).

Anyway, should not block this PR.

```diff
commit 0b9d2657e (HEAD -> cmp-bytes)
Date:   Wed May 15 12:29:14 2019 +0200

    Do not mix unicode/bytes on Python 2

diff --git a/changelog/5267.bugfix.rst b/changelog/5267.bugfix.rst
new file mode 100644
index 000000000..749360195
--- /dev/null
+++ b/changelog/5267.bugfix.rst
@@ -0,0 +1 @@
+Comparisons with assertion rewriting are now skipped with mixed unicode/bytes on Python 2.
diff --git a/src/_pytest/assertion/util.py b/src/_pytest/assertion/util.py
index 23bbe49b7..7e45735e3 100644
--- a/src/_pytest/assertion/util.py
+++ b/src/_pytest/assertion/util.py
@@ -147,9 +147,7 @@ def assertrepr_compare(config, op, left, right):
     explanation = None
     try:
         if op == "==":
-            if isinstance(left, bytes) and isinstance(right, bytes):
-                explanation = _diff_text(left, right, verbose)
-            elif istext(left) and istext(right):
+            if type(left) == type(right) and (isinstance(left, bytes) or istext(left)):
                 explanation = _diff_text(left, right, verbose)
             else:
                 if issequence(left) and issequence(right):
diff --git a/testing/test_assertion.py b/testing/test_assertion.py
index 9f4244640..501bf6ae5 100644
--- a/testing/test_assertion.py
+++ b/testing/test_assertion.py
@@ -361,6 +361,11 @@ def test_bytes_diff(self):
         else:
             assert diff == ["'spam' == 'eggs'", "- spam", "+ eggs"]
 
+    def test_str_and_bytes_diff(self):
+        """pytest_assertrepr_compare does nothing with mixed bytes + string."""
+        diff = callequal(u"spam", b"eggs")
+        assert diff is None
+
     def test_list(self):
         expl = callequal([0, 1], [0, 2])
         assert len(expl) > 1
```